auto_submit/lib/configuration/repository_configuration_manager.dart

```diff
@@ -0,0 +1,174 @@
+// Copyright 2023 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// import 'dart:typed_data';
+
+import 'package:auto_submit/configuration/repository_configuration.dart';
+import 'package:auto_submit/service/config.dart';
+import 'package:auto_submit/service/github_service.dart';
+import 'package:auto_submit/service/log.dart';
+import 'package:github/github.dart';
+import 'package:mutex/mutex.dart';
+import 'package:neat_cache/neat_cache.dart';
+
+/// The [RepositoryConfigurationManager] is responsible for fetching and merging
+/// the autosubmit configuration from the Org level repository and if needed
+/// fetching the override configuration from the pull request repository.
+///
+/// It will attempt to access the cache first before repulling the configuraiton
+/// from the repositories. This is currently set at a 10 minute TTL.
+class RepositoryConfigurationManager {
+  final Mutex _mutex = Mutex();
+
+  static const String fileSeparator = '/';
+  // This is the well named organization level repository and configuration file
+  // we will read before looking to see if there is a local file with
+  // overwrites.
+  static const String orgRepository = '.github';
+  static const String dirName = 'autosubmit';
+  static const String fileName = 'autosubmit.yml';
+
+  final Config config;
+  final Cache cache;
+
+  RepositoryConfigurationManager(this.config, this.cache);
+
+  /// Read the configuration from the cache given the slug, if the config is not
+  /// in the cache then go and get it from the repository and store it in the
+  /// cache.
+  ///
+  /// Entries will be stored in the cache as config/slug/autosubmit.yaml
+  Future<RepositoryConfiguration> readRepositoryConfiguration(
+    RepositorySlug slug,
+  ) async {
+    await _mutex.acquire();
+    try {
+      // Get the contents from the cache or go to github.
+      final cacheValue = await cache['${slug.fullName}$fileSeparator$fileName'].get(
+        () async => _getConfiguration(slug),
+        const Duration(minutes: 10),
```

Comment:
Can we make add this as a constant and document it somewhere? should we add a README file to explain the different configurations?

Reply:
Yeah I can do that.